crates/bifrost/src/providers/replicated_loglet/sequencer/appender.rs

```diff
@@ -149,7 +147,14 @@ impl<T: TransportConnect> SequencerAppender<T> {
                 State::Done | State::Cancelled | State::Sealed => break state,
                 State::Wave { graylist } => {
                     self.current_wave += 1;
-                    let Some(next_state) = cancellation
+                    // # Why is this cancellation safe?
+                    // Because we don't await any futures inside the join_next() loop, so we are
```

Comment:
Why would it be a problem if we awaited futures inside the join_next() loop? If I understand things correctly, then the invariant that must hold is that we only acknowledge the write after we have replicated and updated the tail.

Reply:
I'll try and explain it more in the comment. The invariant you mentioned if correct, but theoretically, if we have an await _after_ updating the global tail, and still marked this append as cancelled, then other appenders after this one _might_ be unblocked and finish their quorum write and therefore reporting success. This means that the writer has a hole in the log. This will not happen with the current one-by-one design, hence the "theoretical" bit.